iwyu_ast_util.cc

```diff
@@ -1422,6 +1432,34 @@ FunctionDecl* GetRedeclSpecifyingDefArg(unsigned i, FunctionDecl* func) {
   CHECK_UNREACHABLE_("There should be a redecl specifying the default arg");
 }
 
+const TemplateDecl* GetRedeclSpecifyingDefArg(unsigned i,
+                                              const TemplateDecl* decl) {
+  const auto* redeclarable = dyn_cast<RedeclarableTemplateDecl>(decl);
+  // If this template isn't redeclarable, then the given declaration should
+  // specify default arguments, if any.
+  if (!redeclarable)
+    return decl;
+  for (const RedeclarableTemplateDecl* redecl : redeclarable->redecls()) {
+    const TemplateParameterList* params = redecl->getTemplateParameters();
+    if (IsDefTplArgSpecified(params->getParam(i)))
+      return redecl;
+  }
+  CHECK_UNREACHABLE_("There should be a redecl specifying the default arg");
+}
+
+const NamedDecl* GetIndependentRedecl(const NamedDecl* decl) {
+  const auto* redeclarable = dyn_cast<RedeclarableTemplateDecl>(decl);
+  if (!redeclarable)
+    return decl;
+  if (const TemplateDecl* found = GetRedeclNotSpecifyingAnyDefArg(redeclarable))
+    return found;
+  // If there is no redeclaration without any explicitly written default
+  // arguments, return the one specifying the last argument because it doesn't
+  // depend on any other redeclaration.
+  unsigned last_param_idx = redeclarable->getTemplateParameters()->size() - 1;
+  return GetRedeclSpecifyingDefArg(last_param_idx, redeclarable);
```

Comment:
Wait, how do we know this doesn't hit the the `CHECK_` above?

Reply:
> that I can't call this function if (num_written_args < tpl->getTemplateParameters()->size()).

You can call it, why not? It doesn't depend on `TemplateSpecializationType` and its written arguments in any way.

It still seems to me that having it a distinct function makes the code more clear.

In fact, ideally, it should not search for the redecl specifying the last default argument but instead somehow construct a new one not specifying any default argument, as the TODO in `VisitTemplateSpecializationType` states. Because if all the template arguments are explicitly written, the template may be forward-declared, and this should not depend on whether there is a redeclaration convenient for IWYU somewhere or not.

Reply:
I can't take it in, too tired :-).

But I understand it better now than before, thank you for explaining!